procrustes/kopt.py

```diff
@@ -34,61 +34,59 @@
+    b : ndarray
+        The 2D-array :math:`\mathbf{B}` representing the reference matrix.
+    p : ndarray, optional
+        The 2D-array :math:`\mathbf{P}` representing the initial permutation matrix. If `None`, the
+        identity matrix is used.
+    k : int, optional
+        The order of the permutation. For example, `k=3` swaps all possible 3-permutations of the
+        given p matrix.
 
     Returns
     -------
     perm : ndarray
         The permutation array after optimal heuristic search.
-    kopt_error : float
+    error : float
         The error distance of two arrays with the updated permutation array.
     """
-    if kopt_k < 2:
-        raise ValueError("Kopt_k value must be a integer greater than 2.")
-    # if perm is not specified, use the identity matrix as default
-    if perm is None:
-        perm = np.identity(np.shape(array_a)[0])
-    num_row = perm.shape[0]
-    kopt_error = ref_error
-    # all the possible row-wise permutations
-    for comb in it.combinations(np.arange(num_row), r=kopt_k):
-        for comb_perm in it.permutations(comb, r=kopt_k):
-            if comb_perm != comb:
-                perm_kopt = deepcopy(perm)
-                perm_kopt[comb, :] = perm_kopt[comb_perm, :]
-                e_kopt_new = compute_error(array_a, array_b, perm_kopt, perm_kopt)
-                if e_kopt_new < kopt_error:
-                    perm = perm_kopt
-                    kopt_error = e_kopt_new
-                    if kopt_error <= kopt_tol:
-                        break
-    return perm, kopt_error
-
-
-def kopt_heuristic_double(array_m, array_n, ref_error,
-                          perm_p=None, perm_q=None,
-                          kopt_k=3, kopt_tol=1.e-8):
+    if k < 2 or not isinstance(k, int):
+        raise ValueError(f"Argument k must be a integer greater than 2. Given k={k}")
+    # assign p to be an identity array, if not specified
+    if p is None:
+        p = np.identity(np.shape(a)[0])
+    # compute 2-sided permutation error of the initial p matrix
+    error = compute_error(a, b, p, p)
+    # swap rows and columns until the permutation matrix is not improved
+    search = True
+    while search:
+        search = False
+        for comb in it.combinations(np.arange(p.shape[0]), r=k):
+            for comb_perm in it.permutations(comb, r=k):
+                if comb_perm != comb:
+                    p_new = deepcopy(p)
+                    p_new[comb, :] = p_new[comb_perm, :]
+                    error_new = compute_error(a, b, p_new, p_new)
+                    if error_new < error:
+                        search = True
+                        p, error = p_new, error_new
+                        # check whether perfect permutation matrix is found
+                        # TODO: smarter threshold based on norm of matrix
+                        if error <= 1.0e-8:
+                            return p, error
+    return p, error
```

Comment:
Does this sound right to you @PaulWAyers & @fwmeng88?

Reply:
Looks good. I would choose the error threshold as something like 1e-8*sqrt(norm(A)*norm(B)). The problem would be revealed by a test matrix that is very ill-scaled, e.g., where A and B are filled with elements of order 1e-30 or something like that.

This isn't *quite* what I'd intended, but it works very well and is even better (more accurate) on average than what I'd intended. The (cheaper) thing I'd intended would break the loops as soon as search=True. This sort of break-a-double-loop issue is something that requires crazy work-arounds in Python. 
    https://note.nkmk.me/en/python-break-nested-loops/

But why do we have a double loop at all? Based on my reading of iter-tools, you can generate all the permutations directly. But that requires a lot more thought to do....not totally sure how to do it.

Reply:
Actually, why make a permutation matrix (mostly zeros and ones) and multiply by A (cost of O(n^3)). You can directly compute A with permuted rows/columns, then compute the error in (permuted) A. This has cost O(n^2) (computing the error) instead of O(n^3). **IF** the error is reduced, you can return the permutation, or even the permutation matrix (which can be constructed at the very end).

Reply:
Good point regarding computation cost. Not worth sticking to using the `compute_error` function, when the cost can be reduced by a factor of O(n). I will implement it.

Reply:
I had some "fun" trying to generate permutations that were inversions up to some order, but that's a HARD problem it seems.
Information Processing Letters; Volume 86, Issue 2, 30 April 2003, Pages 107-112

Reply:
I think you can replace the double-for loop with:
for comb_perm in it.permutations(np.arange(5), r=3):
    comb = tuple(sorted(comb_perm))
    print(comb, comb_perm)
    
There is some other low-hanging fruit here (the same `comb` tuple shows up k! times in a row) but the bottleneck is going to be evaluating the error, so I see no reason to worry about the fact you are sorting a small vector of k entries repeatedly.

Reply:
Great suggestion. I fixed the nested for loops.